sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscriber.java

```diff
@@ -233,20 +260,24 @@ void cancel() {
       if (subscription != null) {
         subscription.cancel();
       }
+      subscription = null;
     }
 
     /**
      * Handles new established {@link Subscription}.
      *
-     * <p>Cancels subscription immediately if {@link EFOShardSubscriber#isStopped} already.
-     * Otherwise, if below the {@link #inFlight} limit, the first event is requested.
+     * <p>Cancels subscription immediately if {@link EFOShardSubscriber#state} is {@link
+     * State#STOPPED} already. Otherwise, if below the {@link #inFlight} limit, the first event is
+     * requested.
      */
     @Override
     public void onSubscribe(Subscription subscription) {
-      if (subscription != null) {
+      this.subscription = subscription;
+      if (state == STOPPED) {
+        cancel();
+      } else if (inFlight.get() < IN_FLIGHT_LIMIT) {
```

Comment:
I think that's where we introduced the bug that caused subscribers to hang... we incremented `inFlight` despite not having received any event.
```
      if (inFlight.incrementAndGet() < IN_FLIGHT_LIMIT && subscription != null) {
        subscription.request(1);
      }
```